### ISSUE TITLE
Quotient semigroups are slow

### ISSUE THREAD

**Issue by ChristopherRussell** [semigroups/Semigroups#456]
This is more of a request. @james-d-mitchell said it would be helpful to have examples like this for targeting future improvements. When I call IsomorphismPermGroup on the quotient semigroup of an inverse semigroup by its minimum group congruence (which is a group) I found this can be very slow for relatively small examples. This is relevant to an improvement I am working on for `IsomorphismSemigroup` (for `McAlsiterTripleSemigroup`'s).

This first example is a semigroup of size 720 (which is a semidirect product of Sym(5) by a semilattice with 6 points where all multiplications x * y = 0 unless x = y != 0). The quotient semigroup by the minimum group congruence is isomorphic to Sym(5) and this calculation hadn't finished after over 5 minutes on my computer.

```
gap> S := InverseSemigroup( [ PartialPerm( [ 1, 2, 3, 4, 5 ], [ 1, 2, 3, 4, 5 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 7 ], [ 2, 5, 1, 4, 3, 6 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 6 ], [ 4, 3, 5, 1, 2, 9 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 10 ], [ 5, 2, 4, 1, 3, 10 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 8 ], [ 1, 5, 4, 2, 3, 7 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 7 ], [ 3, 4, 1, 5, 2, 6 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 10 ], [ 1, 2, 4, 3, 5, 10 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 7 ], [ 2, 1, 4, 3, 5, 9 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 10 ], [ 2, 4, 5, 1, 3, 9 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 9 ], [ 2, 3, 4, 1, 5, 6 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 9 ], [ 2, 1, 3, 4, 5, 9 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 9 ], [ 5, 4, 1, 3, 2, 7 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 7 ], [ 4, 2, 5, 1, 3, 8 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 9 ], [ 4, 1, 3, 2, 5, 10 ] ) ] );
gap> Size(S);
720
gap> cong := MinimumGroupCongruence(S);
<semigroup congruence over <inverse partial perm semigroup of size 720, rank 720 with 720
 generators> with congruence pair (6,1)>
gap> Size(S / cong);
120
gap> IsomorphismPermGroup(S / cong);

```

The next example is a similar example but with a smaller semigroup - the quotient by the minimum group congruence is still isomorphic to Sym(5) but the semigroup is only size 336 as opposed to 720. This finishes in ~3 seconds.

```
gap> S := InverseSemigroup( [ PartialPerm( [ 1, 2, 3, 4, 5 ], [ 1, 2, 3, 4, 5 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 6 ], [ 5, 2, 3, 1, 4, 8 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 6 ], [ 3, 4, 5, 2, 1, 7 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 8 ], [ 3, 1, 4, 2, 5, 8 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 7 ], [ 1, 4, 3, 2, 5, 7 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 7 ], [ 2, 4, 5, 1, 3, 8 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 7 ], [ 1, 2, 5, 4, 3, 8 ] ),
>   PartialPerm( [ 1, 2, 3, 4, 5, 8 ], [ 4, 5, 2, 3, 1, 6 ] ) ] );
gap> Size(S);
336
gap> cong := MinimumGroupCongruence(S);
<semigroup congruence over <inverse partial perm semigroup of size 720, rank 720 with 720
 generators> with congruence pair (6,1)>
gap> Size(S / cong);
120
gap> IsomorphismPermGroup(S / cong);
MappingByFunction( <quotient of <semigroup congruence over <inverse partial perm semigroup of
 rank 8 with 8 generators> with congruence pair (4,1)>>, <group of size 120, with 12 generators>
, function( x ) ... end, function( x ) ... end )
gap> time;
3222
```

**Comment by james-d-mitchell** [comment]
This is related to Issues #370 and #454.

**Comment by wilfwilson** [comment]
Slightly related technical question: if `S` is an inverse semigroup and `cong` is a group congruence, we know as mathematicians that the quotient `S / cong` is a group. How do we use this knowledge in GAP? Should we have a property `IsGroupCongruence` (which is set `true` at creation for a minimum group congruence, for instance). If we do this, then whenever `IsGroupCongruence(cong)` is known be true, and we create the quotient `Parent(cong) / cong`, then we can set the quotient to satisfy `IsGroupAsSemigroup`. Does this sound sensible? Feels like there should be nicer ways of retaining and exploiting this type of information.

**Comment by james-d-mitchell** [comment]
These are all good questions and suggestions. This does not answer your question exactly, but my feeling is that the organisation of the congruences stuff in Semigroups should be reviewed and rewritten. I don't mean the main methods themselves but rather the various categories and what can be expected of the congruences. It evolved naturally and incrementally over time, so at present it is  lots of different things bolted together, a more structured approach would be good. What you're talking about would be part of that. 

In particular, there is not really much support for quotient semigroups in Semigroups (or the GAP library), and I think this would be the first step, and methods for special types of quotients would be part of this. This is connected to the methods for finitely presented semigroups, although not exactly the same. We could open a project for quotient semigroups if you're game @wilfwilson ?

**Comment by wilfwilson** [comment]
I'm not especially game right now, but please feel free to open a project and who knows how I'll feel/what I'll have time for in the future 🙂

**Comment by james-d-mitchell** [comment]
Resolved in v4.0.3
